lib/Service/CategoriesService.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * Calendar App
+ *
+ * @copyright 2023 Claus-Justus Heine <himself@claus-justus-heine.de>
+ *
+ * @author Claus-Justus Heine <himself@claus-justus-heine.de>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Calendar\Service;
+
+use OCP\Calendar\ICalendar;
+use OCP\Calendar\ICalendarQuery;
+use OCP\Calendar\IManager as ICalendarManager;
+use OCP\IL10N;
+use OCP\SystemTag\ISystemTag;
+use OCP\SystemTag\ISystemTagManager;
+use Psr\Log\LoggerInterface;
+
+class CategoriesService {
+	/** @var null|string */
+	private $userId;
+
+	/** @var ICalendarManager */
+	private $calendarManager;
+
+	/** @var ISystemTagManager */
+	private $systemTagManager;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	/** @var IL10N */
+	private $l;
+
+	private const CALENDAR_OBJECT_PROPERTIES_TABLE = 'calendarobjects_props';
+
+	public function __construct(?string $userId,
+		ICalendarManager $calendarManager,
+		ISystemTagManager $systemTagManager,
+		LoggerInterface $logger,
+		IL10N $l10n) {
+		$this->userId = $userId;
+		$this->calendarManager = $calendarManager;
+		$this->systemTagManager = $systemTagManager;
+		$this->logger = $logger;
+		$this->l = $l10n;
+	}
+
+	/**
+	 * This is a simplistic brute-force extraction of all already used
+	 * categories from all events accessible to the currently logged in user.
+	 *
+	 * @return array
+	 */
+	private function getUsedCategories(): array {
+		if (empty($this->userId)) {
+			return [];
+		}
+
+		$categories = [];
+		$principalUri = 'principals/users/' . $this->userId;
+		$query = $this->calendarManager->newQuery($principalUri);
+		$query->addSearchProperty(ICalendarQuery::SEARCH_PROPERTY_CATEGORIES);
+		$calendarObjects = $this->calendarManager->searchForPrincipal($query);
+		foreach ($calendarObjects as $objectInfo) {
+			foreach ($objectInfo['objects'] as $calendarObject) {
+				if (isset($calendarObject['CATEGORIES'])) {
+					$eventCategories = explode(',', $calendarObject['CATEGORIES'][0][0]);
+					$categories = array_merge(
```

Comment:
tip: build a multi-dimensional categories array first and flatten that array after the loop

https://chemaclass.medium.com/never-use-array-merge-in-a-loop-74917b056ff9

Reply:
I am not sure that this applies here. The point is also that we want to have a unique list of categories. Of course, if this was only about merging the individual category array then the repeated array_merge() inside the  loop would not be a good idea.

Of course, I can rephrase that piece, but I am not fully convinced that this will make things faster. A straight forward approach would be to generate a blown up array inside the loop, and then merge it and call array_unique.

Anyhow, see 055cbc7f